services/upgrader/upgrader.go

```diff
@@ -0,0 +1,176 @@
+package upgrader
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"errors"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"runtime"
+	"sync/atomic"
+
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+var upgradePercentage int64
+var upgradeError error
+var upgrading int32
+
+type writeCounter struct {
+	total  int64
+	readed int64
+}
+
+func (wc *writeCounter) Write(p []byte) (int, error) {
+	n := len(p)
+	wc.readed += int64(n)
+	percentage := (wc.readed * 100) / wc.total
+	if percentage == 0 {
+		upgradePercentage = 1
+	} else if percentage == 100 {
+		upgradePercentage = 99
+	} else {
+		upgradePercentage = percentage
+	}
+	return n, nil
+}
+
+func getUpgradeDirectory() (string, error) {
+	exePath, err := os.Executable()
+	if err != nil {
+		return "", err
+	}
+	return filepath.Dir(exePath), nil
+}
+
+func getCurrentVersionTgzUrl() string {
+	return "https://releases.mattermost.com/" + model.CurrentVersion + "/mattermost-" + model.CurrentVersion + "-linux-amd64.tar.gz"
+}
+
+func canIUpgrade() error {
+	if runtime.GOARCH != "amd64" {
```

Comment:
Shouldn't we also check for `GOOS = linux` ?  What happens on mac ?

Reply:
good point, I think this should work in all the operating system, but we should be getting the right build. I'll check it.

Reply:
Ok, I think we only distribute the linux version for the server, so... yes, this feature should be only for linux.

Reply:
Done